### ISSUE TITLE
heap-buffer-overflow read in form_update_line() line=393

### ISSUE THREAD

**Issue by kcwu** [tats/w3m#78]
input (`xxd cases/tats-w3m-78`)
```
00000000: 3c6d 6574 6120 6368 6172 7365 743d 6762  <meta charset=gb
00000010: 3138 3033 303e 0a3c 7465 7874 6172 6561  18030>.<textarea
00000020: 2072 6f77 733d 383e 8093                  rows=8>..
```

how to reproduce:
```
ASAN_OPTIONS=abort_on_error=1:detect_leaks=0 LD_LIBRARY_PATH=./notgc ./w3m-tats.asan -T text/html -dump cases/tats-w3m-78
```

stderr:
```
=================================================================
==3441038==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000d920 at pc 0x00000063345f bp 0x7fffe8197f00 sp 0x7fffe8197ef8
READ of size 1 at 0x60300000d920 thread T0
    #0 0x63345e in form_update_line /targets/w3m-tats/form.c:393:13
    #1 0x630962 in formUpdateBuffer /targets/w3m-tats/form.c:497:12
    #2 0x631c87 in formResetBuffer /targets/w3m-tats/form.c:272:2
    #3 0x55bdd8 in loadHTMLBuffer /targets/w3m-tats/file.c:6800:2
    #4 0x55ee64 in loadSomething /targets/w3m-tats/file.c:224:16
    #5 0x5535ac in loadGeneralFile /targets/w3m-tats/file.c:2241:6
    #6 0x4f9202 in main /targets/w3m-tats/main.c:1017:12
    #7 0x7f2dc837cf44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
    #8 0x41bf25 in _start (/w3m-tats.asan+0x41bf25)

0x60300000d920 is located 0 bytes to the right of 32-byte region [0x60300000d900,0x60300000d920)
allocated by thread T0 here:
    #0 0x4c6288 in __interceptor_malloc (/w3m-tats.asan+0x4c6288)
    #1 0x7f2dc9acac21 in GC_malloc_atomic /notgc/notgc.c:275
    #2 0x573b82 in process_textarea /targets/w3m-tats/file.c:4046:32
    #3 0x58cfb6 in HTMLtagproc1 /targets/w3m-tats/file.c:5154:8
    #4 0x5928ff in HTMLlineproc0 /targets/w3m-tats/file.c:6489:10
    #5 0x5a5d4c in loadHTMLstream /targets/w3m-tats/file.c:7265:2
    #6 0x55bbf8 in loadHTMLBuffer /targets/w3m-tats/file.c:6794:5
    #7 0x55ee64 in loadSomething /targets/w3m-tats/file.c:224:16
    #8 0x5535ac in loadGeneralFile /targets/w3m-tats/file.c:2241:6
    #9 0x4f9202 in main /targets/w3m-tats/main.c:1017:12
    #10 0x7f2dc837cf44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287

SUMMARY: AddressSanitizer: heap-buffer-overflow /targets/w3m-tats/form.c:393:13 in form_update_line
Shadow bytes around the buggy address:
  0x0c067fff9ad0: fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00
  0x0c067fff9ae0: 00 00 fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa
  0x0c067fff9af0: 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00 00 00
  0x0c067fff9b00: fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa fd fd
  0x0c067fff9b10: fd fd fa fa 00 00 00 00 fa fa 00 00 01 fa fa fa
=>0x0c067fff9b20: 00 00 00 00[fa]fa 00 00 00 00 fa fa 00 00 00 00
  0x0c067fff9b30: fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00
  0x0c067fff9b40: 00 00 fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa
  0x0c067fff9b50: 00 00 00 00 fa fa fd fd fd fa fa fa 00 00 00 00
  0x0c067fff9b60: fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00
  0x0c067fff9b70: 00 00 fa fa 00 00 00 fa fa fa fd fd fd fd fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3441038==ABORTING
```

This is detected with help of dummy libgc wrapper. See http://github.com/kcwu/fuzzing-w3m/notgc for detail.
More detail to reproduce please see http://github.com/kcwu/fuzzing-w3m

For your convenience,
gdbline:
LD_LIBRARY_PATH=./notgc ASAN_OPTIONS=abort_on_error=1:detect_leaks=0 gdb --args ./w3m-tats.asan -T text/html -dump cases/tats-w3m-78

This is found by afl-fuzz

**Comment by kcwu** [comment]
I found this is due to incomplete fix of issue #75.
form.c
```
281         char *p, *buf, *q = *str + strlen(*str);
...
329         for (p = *str, w = 0, pos = spos; p < q && w < width;) {
...
332             c_len = get_mclen(p);
...
384             p += c_len;
385         }
...
393                 while (*p && *p != '\r' && *p != '\n')
...
401         if (p > q)
402             p = q;
```
You use `q` to guard pointer `p` at line 401. But line 393 already access beyond the end of string.

**Comment by tats** [comment]
> I found this is due to incomplete fix of issue #75.

Indeed.  Fixed, thank you.

**Comment by tats** [comment]
Patch updated by https://github.com/tats/w3m/commit/6eea841d3a0f8dc539584dc67b15f585a8213775
